### PR TITLE
Internal metrics

### LINKED ISSUE #4006: Create internal monitoring infrastructure
## Context

We want to have some statistics on the internal structures of massa. We can use prometheus exporters to save the content of the internal data and follow their evolution across the time

## TODO

Here is two first things that we can monitor :

- [ ] Create a `massa-channel` module that wrap the crossbeam channel but also export the size of the channel often (when receiving for example or at a certain interval)
- [ ] Monitor the `massa-storage` module to have his number of blocks, ops and endorsements stored displayed often (when reading or at a certain interval)

It would be nice to setup a grafana architecture to view those stats. Maybe we can take example from https://github.com/massalabs/sre-scripts/pull/2

### PR DESCRIPTION
**Pull request by modship**
* [ ] document all added functions
* [x] try in sandbox /simulation/labnet
* [ ] unit tests on the added/changed features
  * [x] make tests compile
  * [x] make tests pass 
* [ ] add logs allowing easy debugging in case the changes caused problems
* [ ] if the API has changed, update the API specification

close #4006 #2404 

TODO : 
 - [x] Create `massa-metrics` package
 - [x] Create `massa-channel` package
 - [x] Add new features **metrics** // REMOVED
 - [x] Start metrics server 
 - [x] Expose metrics in http (for prometheus on `/metrics`)
 - [x] Set listener addr from config
 - [x] tick delay from config
 - [x] Add crossbeam channel wrapper which expose for each channel : 
 	- [X] actual_len 
 	- [X] total_receive 
- [x] Unregister when all ref to the channel are dropped 
 - [x] Counter :
 	- [x] bytes receive / send for peer manager
 	- [x] bytes receive / send for each endpoint 
 	- [x] block graph counter
 	- [x] block graph ms (sum of delta in ms between block inclusion in graph and block slot)
- [x] Gauge : 
     - [x] total blocks in storage
	 - [x] total operations in storage
	 - [X] total endorsements in storage
	- [x] final_cursor_period
	- [x] final_cursor_thread 		
	- [x] active_cursor_period
	- [x] active_cursor_thread 
	- [x] consensus period for each thread
	- [x] block cache (checked_headers_size, blocks_know_by_peer)
	- [x] operation cache (checked_operations, checked_operations_prefix, ops_know_by_peers)
	- [x] consensus state (active_index, active_index_without_ops, incoming_index, discarded_index, block_statuses)
	- [x] endorsement cache (checked_endorsements, know_by_peer)
- [x] Connection metrics
	- [x] IN
	- [x] OUT 
	- [x] bandwidth  (peernet https://github.com/massalabs/PeerNet/pull/56)	
		- [x] total sent / receive by manager
		- [x] total sent / receive for each endpoint  
- [ ] Warning!
	- [x] when the node has no connections with another peer
	- [x] no data sent / receive since x seconds
- [x]  Dashboard

Fix: 

- [x] node panic when error "AlreadyReg" when we try to create a channel that was previously create
- [x] "metrics" feature not trigger in massa-channel

### CODE REVIEW

File: massa-metrics/src/lib.rs [hunk 1]
```diff
@@ -82,14 +98,30 @@ pub struct MassaMetrics {
 
     final_cursor_thread: IntGauge,
     final_cursor_period: IntGauge,
+
+    // peer bandwidth (bytes sent, bytes received)
+    // todo remove Arc<RwLock<...>> ?
+    peers_bandwidth: Arc<RwLock<HashMap<String, (IntCounter, IntCounter)>>>,
```
Comment (modship) [inline massa-metrics/src/lib.rs:104]: Actually i can remove this Arc<RwLock<>> because only the connectivity thread will edit this HashMap

**Comment by damip** [comment]
@modship  alsolet'as avoid magic constants in the code (eg. the scraping interval of 5 seconds) and make it a config file parameter

**Comment by modship** [comment]
> @modship let's avoid magic constants in the code (eg. the scraping interval of 5 seconds) and make it a config file parameter

If there are still hard-coded values, it's an oversight on my part because I created the configuration.
https://github.com/massalabs/massa/pull/4029/files#diff-5275c26ef00940f5cb65a70bbaac1ffe9e40f1fd639b1bec4e7ae9f131972184R224

File: massa-metrics/src/lib.rs [hunk 2]
```diff
@@ -405,6 +415,80 @@ impl MassaMetrics {
...
+
+    pub fn inc_operations_final_counter(&self, diff: u64) {
+        self.operations_final_counter.inc_by(diff);
+    }
+
+    /// Update the bandwidth metrics for all peers
+    /// HashMap<peer_id, (tx, rx)>
+    pub fn update_peers_tx_rx(&self, data: HashMap<String, (u64, u64)>) {
+        if self.enabled {
+            // #[cfg(not(feature = "testing"))]
```
Comment (AurelienFT) [inline massa-metrics/src/lib.rs:435]: dead code

File: massa-protocol-worker/src/connectivity.rs [hunk 3]
```diff
@@ -98,7 +99,7 @@ pub(crate) fn start_connectivity_thread(
             std::thread::sleep(Duration::from_millis(100));
 
             // Create cache outside of the op handler because it could be used by other handlers
-            let total_in_slots = config.peers_categories.values().map(|v| v.max_in_connections_post_handshake).sum::<usize>() + config.default_category_info.max_in_connections_post_handshake + 1;
+            let total_in_slots = config.peers_categories.values().map(|v| v.max_in_connections_post_handshake).sum::<usize>() + config.default_category_info.max_in_connections_post_handshake +1 ;
```
Comment (AurelienFT) [inline massa-protocol-worker/src/connectivity.rs:102]: remove this change

**Review by AurelienFT — APPROVED** [review]
Small changes but the PR is very useful and we already tried it a lot. For a first version of metrics it's really interesting

**Comment by Ben-PH** [comment]
#4148 Might want to look into this.

**Pull request merged**